[DC-979] Refactoring Query Builder

evansuslovich opened the pull request:
- **Not Ready for Review**

I really liked @pshapiro4broad suggestion of creating classes where we could do this: 
```
    Concept concept = new Concept.Builder().build();
    FieldVariable conceptName = concept.name();
    FieldVariable conceptId = concept.concept_id();
    FieldVariable conceptCode = concept.code();
```

instead of this:

```
    TableVariable concept = TableVariable.forPrimary(TablePointer.fromTableName(CONCEPT));
    FieldVariable conceptName = concept.makeFieldVariable(CONCEPT_NAME);
    FieldVariable conceptId = concept.makeFieldVariable(CONCEPT_ID);
    FieldVariable conceptCode = concept.makeFieldVariable(CONCEPT_CODE);
```

What did I do: 
I created the `Concept`, `ConceptAncestor`, `ConceptRelationship`, and `DomainOccurrence` class that allows this for transition. 

Instead of creating custom constructors for these classes, I made a static `Builder` class in `TableVariable` 
___________

Due to all the constructors in the `Query` class, I thought it could be beneficial to implement the builder design pattern. I made a static `Builder` class in `Query` that allows fields to be added and once the query is created, `.build()` can be called to return a `Query` object. 
__________
- Removed constants from `ConceptChildrenQueryBuilder`
- Removed `String rootTableName` from `CriteriaQueryBuilder` since the rootTableName was always "person"

evansuslovich committed:
utilize CONSTANTS

evansuslovich committed:
tableVariable classes

evansuslovich committed:
changing location of imports

evansuslovich committed:
spotless applyg

evansuslovich committed:
added a FilterVariableBuilder object

evansuslovich committed:
created a TableVariableBuilder function which acts as a builder

evansuslovich committed:
using FilterVariableBuilder

evansuslovich committed:
import constants

evansuslovich committed:
remove person root table

evansuslovich committed:
spotless apply

evansuslovich committed:
moving TableVariableBuilder to /query, creating QueryBuilder class

evansuslovich committed:
refactor using QueryBuilder, builder pattern

evansuslovich committed:
spotless apply

evansuslovich committed:
removing HavingFilterVariable

evansuslovich committed:
spotless apply

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Concept.java:
```diff
@@ -0,0 +1,44 @@
+package bio.terra.service.snapshotbuilder.query;
+
+public class Concept extends TableVariable {
```
In general I would favor encapsulation over inheritance, unless it's unworkable due to needing to expose everything in the wrapped object via method forwarding. So here I would instead write:
```
public class Concept {
   private final TableVariable table;
```

and see how much work you can move into this class from the caller

evansuslovich replied:
Having classes like `Concept` and `ConceptAncestor` extend TableVariable allows for them to be treated as TableVariables in the QueryBuilder:

```
    public Builder addTables(List<TableVariable> tables) {
      this.tables = tables;
      return this;
    }
```

and we can then do this

```
  Query createSubQuery(int conceptId) {
    // concept_relationship is primary table for the subquery
    ConceptRelationship conceptRelationship = new ConceptRelationship();
    FieldVariable descendantConceptId = conceptRelationship.concept_id_2();

    return new Query.Builder()
        .select(List.of(descendantConceptId))
        .addTables(List.of(conceptRelationship))
         ...
        .build();
  }

```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Concept.java:
```diff
@@ -0,0 +1,44 @@
+  public static final String CONCEPT_ID = "concept_id";
+  public static final String STANDARD_CONCEPT = "standard_concept";
+  public static final String CONCEPT_CODE = "concept_code";
+  private static final TablePointer tablePointer = TablePointer.fromTableName(TABLE_NAME);
```
I think this works as there's per-object state in `TablePointer` but maybe it's safer to make a new one for each table, unless there's a reason to share them across all tables.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Concept.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Concept extends TableVariable {
+
+  public static final String TABLE_NAME = "concept";
```
I think a goal here would be to try and make these all `private`, it'd be good to look at each one and try and see if that can be done or not.

In the case of field names, using aliases for each result should let us avoid needing to reference them outside this class.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Concept.java:
```diff
@@ -0,0 +1,44 @@
+  }
+
+  public FieldVariable name() {
+    return this.makeFieldVariable(CONCEPT_NAME);
```
It might not be necessary, but I think it would be better to re-use field variables instead of making a new one each time. Using a `Map` as a cache would work well something like
```
    private final Map<String, FieldVariable> fields = new HashMap<>();

  ....

    return fields.computeIfAbsent(CONCEPT_NAME, this::makeFieldVariable);
```

evansuslovich replied:
I like this idea! and it doesn't add much complexity

evansuslovich committed:
create domain occurrence class

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Concept.java:
```diff
@@ -0,0 +1,44 @@
+        tablePointer,
+        tableVariableBuilder.getJoinField(),
+        tableVariableBuilder.getJoinFieldOnParent(),
+        tableVariableBuilder.isLeftJoin());
```
~I don't think we can assume left join here can we?~ nevermind, I see that this is now a boolean flag

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/QueryBuilder.java:
```diff
@@ -0,0 +1,48 @@
+
+import java.util.List;
+
+public class QueryBuilder {
```
Can you make this a static inner class of `Query`, called `Builder`? That is how I'd expect a builder pattern for a class to be written.

evansuslovich replied:
Yes, this was my first implementation but I took the class out for simplicity purposes.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Concept.java:
```diff
@@ -0,0 +1,44 @@
+package bio.terra.service.snapshotbuilder.query;
```
maybe this should be in a subpackage, like `query.omop` or `query.table`?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -46,18 +36,20 @@ public Query buildConceptChildrenQuery(
-    FieldVariable descendantConceptId = conceptAncestor.makeFieldVariable(DESCENDANT_CONCEPT_ID);
+
+    ConceptAncestor conceptAncestor =
+        new ConceptAncestor(
+            new TableVariableBuilder().join(ConceptAncestor.ANCESTOR_CONCEPT_ID).on(conceptId));
```
Can this form can be pushed down to the table class?
```
    ConceptAncestor conceptAncestor = new ConceptAncestor(conceptId);
```
Assuming that we always join through `ConceptAncestor.ANCESTOR_CONCEPT_ID`. If we also join through descendent then maybe we'd need both forms

evansuslovich replied:
It's joined on both `ANCESTOR_CONCEPT_ID` and `DESCENDENT_CONCEPT_ID`. I think i'd rather prefer having it in the builder class to remove from creating more constructors? What are your thoughts on this? 

A possible solution could be making a constructor:

```
  public ConceptAncestor(String joinField, FieldVariable joinOnParent) {
    super(
        tablePointer,
        joinField,
        joinOnParent
        false
        )
  }
  
 // calls would be 
 ConceptAncestor conceptAncestor = new ConceptAncestor(ConceptAncestor.ANCESTOR_CONCEPT_ID, conceptId)
// or 
 ConceptAncestor conceptAncestor = new ConceptAncestor(ConceptAncestor.DESCENDENT_CONCEPT_ID, conceptId)
```

But, this adds more constructors and removes the abstract nature of the builder pattern.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -68,7 +60,9 @@ public Query buildConceptChildrenQuery(
     // COUNT(DISTINCT person_id)
-    FieldVariable count = domainOccurrence.makeFieldVariable(PERSON_ID, "COUNT", "count", true);
+    FieldVariable count =
+        domainOccurrence.makeFieldVariable(
```
I think all occurrence tables would have a `count()` method that does this

evansuslovich replied:
Yes, this is in my next iteration

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/QueryBuilder.java:
```diff
@@ -0,0 +1,48 @@
+
+  public QueryBuilder() {}
+
+  public QueryBuilder addSelect(List<SelectExpression> select) {
```
I would call this `select()` as `add` doesn't add much here IMO
```suggestion
  public QueryBuilder select(List<SelectExpression> select) {
```

evansuslovich committed:
add QueryBuilder to Query class and rename to Builder

evansuslovich committed:
using DomainOccurrence class

evansuslovich committed:
created a new subpackage

evansuslovich committed:
imports and variable names

evansuslovich committed:
cleaning

evansuslovich committed:
imports and changing to use caching for makeVariable

evansuslovich committed:
add Builder class in TableVariableBuilder

evansuslovich committed:
add back final

evansuslovich commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -89,10 +79,15 @@ public Query buildConceptChildrenQuery(
+
+    return new Query.Builder()
+        .select(select)
+        .addTables(tables)
```
remove 'add', not important

s-rubenstein committed:
Add generics builder pattern (#1686)

evansuslovich committed:
add caching

evansuslovich committed:
incorrect syntax for concept creation

evansuslovich committed:
remove add

evansuslovich committed:
add back final

evansuslovich committed:
Merge branch 'develop' into refactor-constants

evansuslovich committed:
gradlew spotless apply

evansuslovich committed:
make TableVariable not Final

evansuslovich committed:
change return response

evansuslovich committed:
adjust tests to support new Builder syntax

evansuslovich commented on src/main/java/bio/terra/service/snapshotbuilder/query/TableVariable.java:
```diff
@@ -1,15 +1,16 @@
 import org.stringtemplate.v4.ST;
 
-public final class TableVariable implements SqlExpression {
+public class TableVariable implements SqlExpression {
```
TableVariable cannot be final because other classes like `Concept` inherit from it

pshapiro4broad replied:
In general I wouldn't make a class `final` unless there's a reason why we'd want to block subclassing it.

evansuslovich committed:
not leftJoining was failing test

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/FieldVariable.java:
```diff
@@ -4,7 +4,7 @@
 import org.stringtemplate.v4.ST;
 
-public class FieldVariable implements SelectExpression {
+public final class FieldVariable implements SelectExpression {
```
Why make this class `final`?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Query.java:
```diff
@@ -22,62 +20,18 @@ public record Query(
-      orderBy = List.of();
-    }
+    groupBy = groupBy != null ? groupBy : List.of();
+    orderBy = orderBy != null ? orderBy : List.of();
```
I think the original form was fine, but if you wanted to rewrite it I would use `Objects.requireNonNullElse()` here:
```suggestion
    groupBy = Objects.requireNonNullElse(groupBy, List.of());
    orderBy = Objects.requireNonNullElse(orderBy, List.of());
```

This avoids having to duplicate the object being tested with its value in the `?:` expression.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/TableVariable.java:
```diff
@@ -82,4 +83,53 @@ public TablePointer getTablePointer() {
     return joinField == null;
   }
+
+  public static class Builder<T extends TableVariable> {
```
I don't think this class needs a builder pattern, there aren't that many arguments so I don't think it merits having the extra code to support it.

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class Concept extends TableVariable {
```
Instead of subclassing, this should use encapsulation
```
public class Concept {
    private final TableVariable tableVariable;
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,59 @@
+    return getFieldVariable(CONCEPT_CODE);
+  }
+
+  public FieldVariable standard_concept() {
```
Even though the underlying field is snake cased, method names in Java are camel cased.
```suggestion
  public FieldVariable standardConcept() {
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,59 @@
+    return getFieldVariable(STANDARD_CONCEPT);
+  }
+
+  public static class Builder extends TableVariable.Builder<Concept> {
```
I don't think these classes need a builder, but if they do, I wouldn't subclass an inner class to do it. Instead make the base class builder a top-level class.

evansuslovich commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -46,36 +33,41 @@ public Query buildConceptChildrenQuery(
-        TableVariable.forJoined(
-            TablePointer.fromTableName(CONCEPT_ANCESTOR), ANCESTOR_CONCEPT_ID, conceptId);
-    FieldVariable descendantConceptId = conceptAncestor.makeFieldVariable(DESCENDANT_CONCEPT_ID);
+
```
```
ConceptAncestor conceptAncestor = ConceptAncestor.joinDescendant(conceptId);
ConceptAncestor conceptAncestor = ConceptAncestor.joinAncestor(conceptId);
```

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java

standard_concept --> standardConcept

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/Query.java

This avoids having to duplicate the object being tested with its value in the ?: expression.

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
remove final

evansuslovich committed:
import Objects, rename standard_concept --> standardConcept

evansuslovich committed:
Removed Builder pattern, updated tests

evansuslovich committed:
Merge branch 'develop' into refactor-constants

evansuslovich committed:
remove     CloudPlatformWrapper platformWrapper = context.getPlatform();

evansuslovich committed:
convert TablePointer to Table

evansuslovich committed:
add fieldVariable functions, create DomainOccurrence on SnapshotBuilderDomainOption

evansuslovich committed:
leftJoinOnDescendantConcept takee in SnapshotBuilderDomainOption

evansuslovich committed:
use domainOption

evansuslovich committed:
Person table

evansuslovich committed:
commented out a test, will update soon, import Table to convert TableVariable

evansuslovich committed:
use domainOption

evansuslovich committed:
use domainOption

evansuslovich committed:
rename import

evansuslovich committed:
clean up rootTable

evansuslovich committed:
use domainOption

evansuslovich committed:
use SnapshotBuilderDomainOption in leftJoinOnDescendantConcept

evansuslovich committed:
testing tables, cleaning up DomainOccurrence

evansuslovich committed:
adding ConceptAncestor argument in leftJoinOnDescendantConceptto remove the need of instantiating another ConceptAncestor on each time this is called, add another getFieldVariable that creates alias counts

evansuslovich committed:
make forJoined public, remove Builder

evansuslovich committed:
make CriteriaQueryBuilderTest class public so I can access assertQueryEquals

evansuslovich committed:
add test org

evansuslovich committed:
testing

evansuslovich committed:
refactor DomainOccurrence to take in DomainOption and TablePointer

evansuslovich committed:
add fieldVariable in leftJoinOn, call countPersonId

evansuslovich committed:
add fieldVariable in leftJoinOn, call countPersonId

evansuslovich committed:
use static forPrimary

evansuslovich committed:
add fieldVariable in leftJoinOn, call countPersonId

evansuslovich committed:
refactor to camelCase

evansuslovich committed:
fixing merge conflicts

evansuslovich committed:
removed, unnecessary

evansuslovich committed:
remove descendantConceptId

evansuslovich committed:
add private constructor

evansuslovich committed:
left comment in Query constructor

evansuslovich committed:
add variable for option

evansuslovich committed:
Table implements SqlExpression

evansuslovich committed:
slight refactor

evansuslovich committed:
remove unnecessary test

evansuslovich committed:
gradlew SpotlessApply

evansuslovich committed:
remove public modifier

evansuslovich committed:
calling the QueryBuilderFactory to cover this line of code

evansuslovich committed:
Manually resolved merge conflicts

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/Query.java:
```diff
@@ -151,7 +103,55 @@ public String renderSQL(SqlRenderContext context) {
+
+    public Query build() {
+      return new Query(
+          this.select, this.tables, this.where, this.groupBy, this.orderBy, this.limit);
```
In general, the TDR code only uses `this.` in places where it's required for disambiguation purposes, like a constructor or setter. So to use the TDR convention:
```suggestion
          select, tables, where, groupBy, orderBy, limit);
```

rjohanek replied:
nice!

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/SourceVariable.java:
```diff
@@ -50,7 +50,7 @@ public static SourceVariable forLeftJoined(
   }
 
-  private static SourceVariable forJoined(
+  public static SourceVariable forJoined(
```
The only use of this outside the class is in `DomainOccurrence.leftJoinOn()`, could that call `forLeftJoined()` instead, and keep this private?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/TableVariable.java:
```diff
@@ -0,0 +1 @@
+
```
This should be deleted

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,52 @@
+    return getFieldVariable(DOMAIN_ID);
+  }
+
+  public FieldVariable code() {
```
Maybe this should be `conceptCode()`, to match the other method names here?
```suggestion
  public FieldVariable conceptCode() {
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,52 @@
+    super(SourceVariable.forJoined(tablePointer, joinField, joinFieldOnParent));
+  }
+
+  public static Concept asPrimary() {
```
Maybe this should be `forPrimary()`, to match `SourceVariable.forPrimary()`

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,52 @@
+    return new Concept();
+  }
+
+  public static Concept joinConceptId(FieldVariable joinFieldOnParent) {
```
Similarly, `SourceVariable` has `forJoined()`, so maybe this should follow that convention too? Something like `forJoinedConceptId()`?

Either that, or change `SourceVariable`'s method names to match this convention, maybe `joinOn()`?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java:
```diff
@@ -0,0 +1,52 @@
+  public static final String CONCEPT_ID = "concept_id";
+  public static final String STANDARD_CONCEPT = "standard_concept";
+  public static final String CONCEPT_CODE = "concept_code";
+  private static final TablePointer tablePointer = TablePointer.fromTableName(TABLE_NAME);
```
I'm still not sure if this should be a `static` or not. It's immutable so there's no harm in it, but it's also small and there's not much overhead in making a new one each time. But if it is `static`, then it should be renamed to follow the java naming conventions
```suggestion
  private static final TablePointer TABLE_POINTER = TablePointer.fromTableName(TABLE_NAME);
```

or maybe
```suggestion
  private static final TablePointer POINTER = TablePointer.fromTableName(TABLE_NAME);
```

evansuslovich committed:
add requireStandardConcept function

evansuslovich committed:
added static selectHasChildren function

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/DomainOccurrence.java:
```diff
@@ -0,0 +1,49 @@
+  private DomainOccurrence(SnapshotBuilderDomainOption domainOption, SourceVariable tableVariable) {
+    super(tableVariable);
+    this.domainOption = domainOption;
+    this.personCount = tableVariable.makeFieldVariable(PERSON_ID, "COUNT", "count", true);
```
Instead of `Person.PERSON_ID`, this should define its own `PERSON_ID` field, since it's referring to the column in the domain occurrence table, not the one in the person table.

For `count`, this should use the alias name from `QueryBuilderFactory` since this is creating an alias that the result mapping code will need to use, so the two names need to agree.
```suggestion
    this.personCount = tableVariable.makeFieldVariable(PERSON_ID, "COUNT", QueryBuilderFactory.COUNT, true);
```

evansuslovich committed:
call static selectHasChildren from ConceptAncestor

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/DomainOccurrence.java:
```diff
@@ -0,0 +1,49 @@
+    this.personCount = tableVariable.makeFieldVariable(PERSON_ID, "COUNT", "count", true);
+  }
+
+  public static DomainOccurrence leftJoinOn(
```
This should also be renamed to match `SourceVariable.forLeftJoined()` (or that should be renamed to match this one).

evansuslovich committed:
call ConceptAncestor selectHasChildren and Concept, requireStandardConcept

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/DomainOccurrence.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public FieldVariable getJoinColumn() {
+    return getFieldVariable(this.domainOption.getColumnName());
```
unnecessary `this.`
```suggestion
    return getFieldVariable(domainOption.getColumnName());
```

evansuslovich committed:
call refactored query code

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/DomainOccurrence.java:
```diff
@@ -0,0 +1,49 @@
+    return getFieldVariable(PERSON_ID);
+  }
+
+  public FieldVariable countPersonId() {
```
Maybe this should be `getCountPersonId()`?

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Person.java:
```diff
@@ -0,0 +1,52 @@
+    super(sourceVariable);
+    this.count =
+        new FieldVariable(
+            new FieldPointer(tablePointer, PERSON_ID, "COUNT"), sourceVariable, null, true);
```
Instead of referring to `tablePointer` directly I think this should use the copy from `sourceVariable`
```suggestion
            new FieldPointer(sourceVariable.getSourcePointer(), PERSON_ID, "COUNT"), sourceVariable, null, true);
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Person.java:
```diff
@@ -0,0 +1,52 @@
+  private static final TablePointer tablePointer = TablePointer.fromTableName(TABLE_NAME);
+  private final FieldVariable count;
+
+  public Person(SourceVariable sourceVariable) {
```
I think this can be `private`
```suggestion
  private Person(SourceVariable sourceVariable) {
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Person.java:
```diff
@@ -0,0 +1,52 @@
+    return new Person(SourceVariable.forPrimary(tablePointer));
+  }
+
+  public FieldVariable fromColumn(String columnName) {
```
I think this and the other methods that return `FieldVariable` should be prefixed with `get`, to match the other table classes
```suggestion
  public FieldVariable getColumn(String columnName) {
```

Also, it looks like this method is only used from a test, if so, can it be removed?

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/query/Query.java:
```diff
@@ -151,7 +103,55 @@ public String renderSQL(SqlRenderContext context) {
 
-  public SourceVariable getPrimaryTable() {
-    return tables.stream().filter(SourceVariable::isPrimary).findFirst().orElseThrow();
+  public Table getPrimaryTable() {
```
Why is this table and not sourceVariable?

evansuslovich replied:
Shelby renamed from `TableVariable` to `SourceVariable`. Before merging with Shelby, I renamed the `TableVariable` to `Table` because the tables were no longer going to be a general `TableVariable` but a more specific `Table` type, which are `Concept`, `ConceptAncestor` ... `Person`. So, we don't want to return a general `SourceVariable` because its encapsulated inside of `Table`.

evansuslovich replied:
We can ask a `Concept` object, "hey, are you primary" like this: 
```
Concept concept = Concept.asPrimary() // creates a concept object
boolean isPrimary = concept.isPrimary(); 
```

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/query/filtervariable/HavingFilterVariable.java:
```diff
@@ -1,20 +0,0 @@
-
-  @Override
-  public String renderSQL(SqlRenderContext context) {
-    return String.format("HAVING COUNT(*) %s %s", operator.renderSQL(context), value);
```
Why was this deleted? Is it replaced or just unused?

evansuslovich replied:
Unused!

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/DomainOccurrence.java:
```diff
@@ -0,0 +1,49 @@
+  }
+
+  public FieldVariable getPerson() {
+    return getFieldVariable(PERSON_ID);
```
just for consistency of naming across these new table files - here you use getPerson, while in the other files you don't use "get" at all. Maybe this could be "personId()" instead

evansuslovich replied:
Perfect! thank you

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/DomainOccurrence.java:
```diff
@@ -0,0 +1,49 @@
+    return getFieldVariable(PERSON_ID);
+  }
+
+  public FieldVariable countPersonId() {
```
is it an ID? Or just a count - personCount() ?

evansuslovich replied:
Ya, I guess you're right, this is the sql it generates in SearchConceptsQueryBuilder: 

```
COUNT(DISTINCT o.person_id) AS count,
```

so it counts the distinct id's of persons?

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Person.java:
```diff
@@ -0,0 +1,52 @@
+    return getFieldVariable(rangeCriteria.getColumnName());
+  }
+
+  public FieldVariable countPersonId() {
```
same thing here - is it an ID? Or just a count ?

evansuslovich replied:
Just count.

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/ConceptChildrenQueryBuilder.java:
```diff
@@ -13,105 +12,102 @@
+    Concept concept = Concept.asPrimary();
+    FieldVariable conceptName = concept.name();
+    FieldVariable conceptId = concept.conceptId();
+    FieldVariable conceptCode = concept.code();
```
looks good!

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/ConceptAncestor.java:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  public static FieldVariable selectHasChildren(SourceVariable joinHasChildren) {
+    return joinHasChildren.makeFieldVariable(DESCENDANT_CONCEPT_ID, "COUNT", "has_children", true);
```
This should the constant from `QueryBuilderFactory` 
```suggestion
    return joinHasChildren.makeFieldVariable(DESCENDANT_CONCEPT_ID, "COUNT", QueryBuilderFactory.HAS_CHILDREN`, true);
```

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/ConceptRelationship.java:
```diff
@@ -0,0 +1,34 @@
+    return new ConceptRelationship();
+  }
+
+  public FieldVariable conceptId1() {
```
To follow convention, these APIs should begin with `get`
```suggestion
  public FieldVariable getConceptId1() {
```

rjohanek commented on src/main/java/bio/terra/service/snapshotbuilder/utils/constants/ConditionOccurrence.java:
```diff
@@ -1,6 +1,7 @@
 
-public final class ConditionOccurrence {
+public class ConditionOccurrence {
+
```
why are these tables (ConditionOccurrence and ObservationOccurrence) not moved over to the tables directory as well? Separate question, If we still need these tables, what is the purpose of the new DomainOccurrence table? Do we need tables for each domain - procedure, device, etc?

evansuslovich committed:
rename from getPersonId to personId and countPersonId to countPerson

rjohanek commented on src/test/java/bio/terra/service/snapshotbuilder/query/QueryTest.java:
```diff
@@ -70,96 +68,11 @@ void renderSqlGroupBy(SqlRenderContext context) {
-  @ParameterizedTest
-  @ArgumentsSource(SqlRenderContextProvider.class)
-  void renderComplexSQL(SqlRenderContext context) {
-    TablePointer tablePointer = TablePointer.fromTableName(Person.TABLE_NAME);
```
why do we no longer need this test?

evansuslovich replied:
This doesn't add any test coverage, everything it's testing has already been tested.

evansuslovich committed:
typo

evansuslovich committed:
renamings

rjohanek reviewed (commented):
the query builder code looks great! very sleek. I just have some questions for my own understanding

pshapiro4broad commented on src/main/java/bio/terra/service/snapshotbuilder/query/tables/Table.java:
```diff
@@ -0,0 +1,30 @@
+package bio.terra.service.snapshotbuilder.query.tables;
```
I wonder if this should be `table` instead? Package names are usually singular 
```suggestion
package bio.terra.service.snapshotbuilder.query.table;
```

pshapiro4broad commented on src/test/java/bio/terra/service/auth/iam/IamServiceTest.java:
```diff
@@ -92,6 +92,23 @@ void testDeletePolicyMember() throws InterruptedException {
   }
 
+  @Test
+  void testDeletePolicyMemberAggregateDataReader() throws InterruptedException {
```
I think this was a merge error, this is the same as `testDeletePolicyMember()` above. Maybe this method was renamed?

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/tables/ConceptAncestor.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
renamed package name tables to table

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/tables/ConceptRelationship.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/tables/Person.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/tables/Concept.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
Merge branch 'refactor-constants' of https://github.com/DataBiosphere/jade-data-repo into refactor-constants

evansuslovich committed:
refactor

evansuslovich committed:
Delete TableVariable

evansuslovich committed:
Update src/main/java/bio/terra/service/snapshotbuilder/query/Query.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
spotlessApply

evansuslovich committed:
make forJoined private

pshapiro4broad commented on src/test/java/bio/terra/service/snapshotbuilder/query/table/PersonTest.java:
```diff
@@ -0,0 +1,43 @@
+  @ParameterizedTest
+  @ArgumentsSource(SqlRenderContextProvider.class)
+  void testCountPersonId(SqlRenderContext context) {
+    FieldVariable fieldVariable = Person.forPrimary().countPerson();
+    assertQueryEquals("COUNT(DISTINCT p.person_id)", fieldVariable.renderSQL(context));
```
This is probably fine but it does duplicate testing that's done elsewhere. The only code to be tested in `Person` is
```
    super(sourceVariable);
    this.countPerson = sourceVariable.makeFieldVariable(PERSON_ID, "COUNT", null, true);
```
and the getter for `countPerson()`.

The duplication isn't good, but a bigger problem is that it makes this test depend on behavior of code outside `Person`. One way to fix this is to write the test using a mocked `SourceVariable` and not using `renderSql()`. For example

```
  @Test
  void testCountPersonId() {
    SourceVariable sourceVariable = mock(SourceVariable.class);
    FieldVariable count = new FieldVariable(null, null);
    when(sourceVariable.makeFieldVariable("person_id", "COUNT", null, true)).thenReturn(count);
    Person person = new Person(sourceVariable);
    assertThat(person.countPerson(), is(count));
  }
```

Right now we do use `renderSQL()` a lot so maybe this PR isn't the right time to start writing tests differently, but I think we should start thinking of how to get away from using it for all tests.

pshapiro4broad commented on src/test/java/bio/terra/service/snapshotbuilder/query/table/PersonTest.java:
```diff
@@ -0,0 +1,43 @@
+  @ParameterizedTest
+  @ArgumentsSource(SqlRenderContextProvider.class)
+  void testAsPrimary(SqlRenderContext context) {
+    Person person = Person.forPrimary();
+    assertQueryEquals("person AS p", person.renderSQL(context));
```
Should also assert that `person.getPrimary()` returns `true`

pshapiro4broad commented on src/test/java/bio/terra/service/snapshotbuilder/query/table/DomainOccurrenceTest.java:
```diff
@@ -0,0 +1,61 @@
+    SnapshotBuilderDomainOption domainOption = new SnapshotBuilderDomainOption();
+    domainOption
+        .root(new SnapshotBuilderConcept())
+        .tableName(ConditionOccurrence.TABLE_NAME)
```
These names don't have meaning outside this test so it can be changed to declare them locally inside this test class, or this test method.

pshapiro4broad commented on src/test/java/bio/terra/service/snapshotbuilder/query/table/ConceptTest.java:
```diff
@@ -0,0 +1,50 @@
+    Concept concept = Concept.asPrimary();
+
+    Map<FieldVariable, String> fieldVariablesToResult =
+        Map.ofEntries(
+            Map.entry(concept.name(), "c.concept_name"),
+            Map.entry(concept.conceptId(), "c.concept_id"),
+            Map.entry(concept.domainId(), "c.domain_id"),
+            Map.entry(concept.conceptCode(), "c.concept_code"),
+            Map.entry(concept.standardConcept(), "c.standard_concept"));
```
`Map.entry()` isn't needed here, you can use `Map.of()` directly
```suggestion
        Map.of(
            concept.name(), "c.concept_name",
            concept.conceptId(), "c.concept_id",
            concept.domainId(), "c.domain_id",
            concept.conceptCode(), "c.concept_code",
            concept.standardConcept(), "c.standard_concept");
```

evansuslovich committed:
moving constants folder into test, only used in testing folders

pshapiro4broad commented on src/test/java/bio/terra/service/snapshotbuilder/SnapshotBuilderServiceTest.java:
```diff
@@ -426,7 +425,7 @@ void testParentQueryResult() throws Exception {
 
   @Test
-  void testRejectRequest() {
+  void testRejectSnapshotAccessRequest() {
```
I think this is a lost merge update from `develop` and should be reverted
```suggestion
  void testRejectRequest() {
```

pshapiro4broad reviewed (approved):
A few more comments, all minor things, looks OK overall

evansuslovich committed:
Update src/test/java/bio/terra/service/snapshotbuilder/SnapshotBuilderServiceTest.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
Update src/test/java/bio/terra/service/snapshotbuilder/query/table/ConceptTest.java

Co-authored-by: Phil Shapiro <pshapiro@broadinstitute.org>

evansuslovich committed:
remove map.ofEntries

evansuslovich committed:
mocked test

evansuslovich committed:
remove test, duplicated code

evansuslovich committed:
gradlew spotless apply, last commit, thank you everyone for the help
